File: app/src/main/java/com/andb/apps/corners/CornerService.java
package com.andb.apps.corners;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.content.res.Configuration;
import android.graphics.Color;
import android.graphics.PixelFormat;
import android.graphics.Point;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffColorFilter;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.os.IBinder;
import android.provider.Settings;
import android.util.Log;
import android.view.Display;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.WindowManager;
import android.widget.TextView;
import android.widget.Toast;

import java.lang.reflect.InvocationTargetException;


public class CornerService extends Service {

    static View mView;

    public LayoutInflater inflater;
    public static WindowManager windowManager;
    static WindowManager.LayoutParams params;

    public static int size;

    Notification notification;
    private NotificationManager notifManager;

    static boolean first;

    public static int NOTIFICATION_ID = 1;
    public static String channelId = "default_channel_id";


    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onCreate() {
        super.onCreate();


        first = false;

        Intent notificationIntent = new Intent(this, MainActivity.class);
        PendingIntent pendingIntent =
                PendingIntent.getActivity(this, 0, notificationIntent, 0);

        if (notifManager == null) {
            notifManager =
                    (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {

            String channelDescription = "Corners";
            //Check if notification channel exists and if not create one
            NotificationChannel notificationChannel = notifManager.getNotificationChannel(channelId);
            if (notificationChannel == null) {
                int importance = NotificationManager.IMPORTANCE_DEFAULT;
                notificationChannel = new NotificationChannel(channelId, channelDescription, importance);
                notificationChannel.setLightColor(Color.BLACK);
                notificationChannel.enableVibration(true);
                notifManager.createNotificationChannel(notificationChannel);
                first = true;
            }


            notification = new Notification.Builder(this, channelId)
                    .setContentTitle(getText(R.string.notification_title))
                    .setContentText(getText(R.string.notification_text))
                    .setSmallIcon(R.drawable.ic_notif)
                    .setContentIntent(pendingIntent)
                    .build();
        } else {
            notification = new Notification.Builder(this)
                    .setContentTitle(getText(R.string.notification_title))
                    .setContentText(getText(R.string.notification_text))
                    .setContentIntent(pendingIntent)
                    .build();
        }


        startForeground(NOTIFICATION_ID, notification);

        Log.d("serviceStart", "service started");

        size = MainActivityFragment.size;




        /*WindowManager.LayoutParams params = new WindowManager.LayoutParams(
                WindowManager.LayoutParams.MATCH_PARENT,
                WindowManager.LayoutParams.MATCH_PARENT,
                LAYOUT_FLAG,
                WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE | WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN | WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE| WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL,
                PixelFormat.TRANSLUCENT);
        params.gravity = Gravity.START | Gravity.TOP;*/


        windowManager = (WindowManager) getSystemService(WINDOW_SERVICE);

        inflater = LayoutInflater.from(CornerService.this);

        mView = inflater.inflate(R.layout.overlay, null);
        setSize(getApplicationContext());
        setColor(MainActivityFragment.cornerColor);

        Point screenSize = getRealScreenSize(this);


        int height = screenSize.y;
        int width = screenSize.x;


        if(Build.VERSION.SDK_INT>=Build.VERSION_CODES.M){
            if (Settings.canDrawOverlays(getApplicationContext())) {
                startOverlay(height, width);
            }else {
                Toast.makeText(getApplicationContext(), R.string.permission_service_sync_error_text, Toast.LENGTH_LONG).show();
            }
        }else {
            startOverlay(height, width);
        }

    }

    public static void startOverlay(int height, int width) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            params = new WindowManager.LayoutParams(width, height

                    /*ViewGroup.LayoutParams.MATCH_PARENT*/, 0, 0,
                    WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY,
                    WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE
                            | WindowManager.LayoutParams.FLAG_LAYOUT_INSET_DECOR
                            | WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED
                            | WindowManager.LayoutParams.FLAG_FULLSCREEN
                            | WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS
                            | WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN
                            | WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE
                            | WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL,
                    PixelFormat.TRANSLUCENT);
        } else {
            params = new WindowManager.LayoutParams(width,
                    height, 0, 0
                    , WindowManager.LayoutParams.TYPE_PRIORITY_PHONE,
                    WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE
                            | WindowManager.LayoutParams.FLAG_LAYOUT_INSET_DECOR
                            | WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED
                            | WindowManager.LayoutParams.FLAG_FULLSCREEN
                            | WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS
                            | WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN
                            | WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE
                            | WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL,
                    PixelFormat.TRANSLUCENT);
        }

        params.gravity = Gravity.TOP | Gravity.START;
        params.x = 0;
        params.y = 0;

        windowManager.addView(mView, params);

        MainActivityFragment.setIndividualVisibility();
        setColor(MainActivityFragment.cornerColor);


        Log.d("popupWindowService", Boolean.toString(first));

        if (first) {
            MainActivityFragment.showTutorial();

            first = false;

        }
    }


    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);

        Point screenSize = getRealScreenSize(this);

        try {
            windowManager.removeViewImmediate(mView);
        }catch (Exception e){
            Toast.makeText(getApplicationContext(), R.string.permission_service_sync_error_text, Toast.LENGTH_LONG).show();
            e.printStackTrace();
        }


        int height = screenSize.y;
        int width = screenSize.x;

        Log.d("configchange", "config changed, restarting overlay");


        if(Build.VERSION.SDK_INT>=Build.VERSION_CODES.M){
            if (Settings.canDrawOverlays(getApplicationContext())) {
                startOverlay(height, width);
            }
        }else {
            startOverlay(height, width);
        }

    }


    public static Point getRealScreenSize(Context context) {
        WindowManager windowManager = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
        Display display = windowManager.getDefaultDisplay();
        Point size = new Point();

            display.getRealSize(size);


        return size;
    }

    public static void setSize(Context context) {


        if(mView!=null) {
            final float scale = context.getResources().getDisplayMetrics().density;

            int dpAdjust = (int) (size * scale + 0.5f);

            TextView topLeft = mView.findViewById(R.id.topLeft);
            TextView topRight = mView.findViewById(R.id.topRight);
            TextView bottomLeft = mView.findViewById(R.id.bottomLeft);
            TextView bottomRight = mView.findViewById(R.id.bottomRight);

            topLeft.setWidth(dpAdjust);
            topLeft.setHeight(dpAdjust);
            topRight.setWidth(dpAdjust);
            topRight.setHeight(dpAdjust);
            bottomLeft.setWidth(dpAdjust);
            bottomLeft.setHeight(dpAdjust);
            bottomRight.setWidth(dpAdjust);
            bottomRight.setHeight(dpAdjust);
        }

    }

    public static void setColor(int color){

        Log.d("setColor" , "setting color: " + Integer.toHexString(color));
        if(mView!=null) {

            TextView topLeft = mView.findViewById(R.id.topLeft);
            TextView topRight = mView.findViewById(R.id.topRight);
            TextView bottomLeft = mView.findViewById(R.id.bottomLeft);
            TextView bottomRight = mView.findViewById(R.id.bottomRight);


            topLeft.getBackground().mutate().setColorFilter(color, PorterDuff.Mode.SRC_IN);
            topRight.getBackground().mutate().setColorFilter(color, PorterDuff.Mode.SRC_IN);
            bottomLeft.getBackground().mutate().setColorFilter(color, PorterDuff.Mode.SRC_IN);
            bottomRight.getBackground().mutate().setColorFilter(color, PorterDuff.Mode.SRC_IN);


            Log.d("setColor", "done setting color");
        }

    }


    @Override
    public void onDestroy() {

        //bad code but prevents settings errors that I think might happen

        try {
            windowManager.removeViewImmediate(mView);
        }catch (Exception e){
            e.printStackTrace();
        }
    }
}

File: app/src/main/java/com/andb/apps/corners/CornerService.kt
package com.andb.apps.corners

import android.annotation.SuppressLint
import android.app.*
import android.content.Context
import android.content.Intent
import android.content.res.Configuration
import android.graphics.PixelFormat
import android.graphics.PorterDuff
import android.os.Build
import android.os.IBinder
import android.provider.Settings
import android.util.Log
import android.view.*
import android.widget.TextView
import android.widget.Toast
import kotlinx.android.synthetic.main.overlay.view.*


const val NOTIFICATION_ID = 1

class CornerService : Service() {

    private var notifManager: NotificationManager? = null

    private lateinit var mView: View
    private lateinit var windowManager: WindowManager


    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    @SuppressLint("InflateParams")
    override fun onCreate() {
        super.onCreate()
        cornerService = this

        if (notifManager == null) {
            notifManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        }

        val notifBuilder: Notification.Builder
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            //Check if notification channel exists and if not create one
            var notificationChannel: NotificationChannel? = notifManager!!.getNotificationChannel(channelId)
            if (notificationChannel == null) {
                notificationChannel = createNotificationChannel()
                notifManager!!.createNotificationChannel(notificationChannel)
            }
            notifBuilder = Notification.Builder(this, channelId)

        } else {
            @Suppress("DEPRECATION")
            notifBuilder = Notification.Builder(this)
        }

        val notificationIntent = Intent(this, MainActivity::class.java)
        val pendingIntent = PendingIntent.getActivity(this, 0, notificationIntent, 0)

        val notification = notifBuilder.setContentTitle(getText(R.string.notification_title))
            .setContentText(getText(R.string.notification_text))
            .setSmallIcon(R.drawable.app_icon_foreground)
            .setContentIntent(pendingIntent)
            .build()

        startForeground(NOTIFICATION_ID, notification)

        Log.d("serviceStart", "service started")


        windowManager = getSystemService(Context.WINDOW_SERVICE) as WindowManager

        val inflater = LayoutInflater.from(this@CornerService)

        mView = inflater.inflate(R.layout.overlay, null)


        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (Settings.canDrawOverlays(applicationContext)) {
                refreshOverlay()
            } else {
                Toast.makeText(applicationContext, R.string.permission_service_sync_error_text, Toast.LENGTH_LONG).show()
            }
        } else {
            refreshOverlay()
        }

    }


    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)


        try {
            windowManager.removeViewImmediate(mView)
        } catch (e: Exception) {
            Toast.makeText(applicationContext, R.string.permission_service_sync_error_text, Toast.LENGTH_LONG)
                .show()
            e.printStackTrace()
        }




        Log.d("configChange", "config changed, restarting overlay")


        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (Settings.canDrawOverlays(applicationContext)) {
                refreshOverlay()
            }
        } else {
            refreshOverlay()
        }

    }


    override fun onDestroy() {

        //bad code but prevents settings errors that I think might happen
        try {
            windowManager.removeViewImmediate(mView)
        } catch (e: Exception) {
            e.printStackTrace()
        }

        cornerService = null
    }


    fun refreshOverlay() {
        val screenSize = getRealScreenSize(this)

        val height = screenSize.y
        val width = screenSize.x

        Log.d("cwNavBar", "${(landscapeFixDefault() xor Values.landscapeFix) && windowManager.defaultDisplay.rotation == Surface.ROTATION_270}")

        val params = getParams(width, height)

        params.gravity = Gravity.TOP or Gravity.START

        //when the navigation bar is on the left, the x is at the edge of the bar, not the edge of the screen, so it needs to be moved back
        params.x = if ((landscapeFixDefault() xor Values.landscapeFix) && windowManager.defaultDisplay.rotation == Surface.ROTATION_270) -getNavigationBarSize(applicationContext).x else 0
        params.y = 0

        setCorners()
        if(mView.parent == null){
            windowManager.addView(mView, params)
        }else{
            windowManager.updateViewLayout(mView, params)
        }

    }

    private fun setCorners() {
        val cornerViews: List<TextView> = with(mView) {
            listOf(topLeft, topRight, bottomLeft, bottomRight)
        }
        cornerViews.forEachIndexed { i, view ->
            val corner = Values.corners[i]

            val scale = view.context.resources.displayMetrics.density
            val size = (corner.size * scale + 0.5f).toInt()

            view.width = size
            view.height = size
            view.visibility = if (corner.visible) View.VISIBLE else View.GONE
            view.background.mutate().setColorFilter(corner.color, PorterDuff.Mode.SRC_ATOP)
        }
    }

    private fun getParams(width: Int, height: Int): WindowManager.LayoutParams {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            return WindowManager.LayoutParams(
                width, height, 0, 0,
                WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY,
                WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE
                        or WindowManager.LayoutParams.FLAG_LAYOUT_INSET_DECOR
                        or WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED
                        or WindowManager.LayoutParams.FLAG_FULLSCREEN
                        or WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS
                        or WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN
                        or WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE
                        or WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL
                        or WindowManager.LayoutParams.FLAG_TRANSLUCENT_NAVIGATION,
                PixelFormat.TRANSLUCENT
            )
        } else {
            return WindowManager.LayoutParams(
                width, height, 0, 0, WindowManager.LayoutParams.TYPE_PRIORITY_PHONE,
                WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE
                        or WindowManager.LayoutParams.FLAG_LAYOUT_INSET_DECOR
                        or WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED
                        or WindowManager.LayoutParams.FLAG_FULLSCREEN
                        or WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS
                        or WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN
                        or WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE
                        or WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL,
                PixelFormat.TRANSLUCENT
            )
        }
    }

    companion object {
        var cornerService: CornerService? = null
    }

}


